<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings SingleCellExperiment to the Tidyverse  • tidySingleCellExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings SingleCellExperiment to the Tidyverse ">
<meta property="og:description" content="tidySingleCellExperiment is an adapter that abstracts
    the SingleCellExperiment container in the form of a tibble.
    This allows *tidy* data manipulation, nesting, and plotting.
    For example, a tidySingleCellExperiment is directly compatible
    with functions from tidyverse packages `dplyr` and `tidyr`,
    as well as plotting with `ggplot2` and `plotly`.
    In addition, the package provides various utility
    functions specific to single-cell omics data analysis
    (e.g., aggregation of cell-level data to pseudobulks).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidySingleCellExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.13.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidySingleCellExperiment package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidySingleCellExperiment/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="tidysinglecellexperiment---part-of-tidytranscriptomics">tidySingleCellExperiment - part of tidytranscriptomics<a class="anchor" aria-label="anchor" href="#tidysinglecellexperiment---part-of-tidytranscriptomics"></a>
</h1></div>
<!-- badges: start -->

<p><strong>Brings SingleCellExperiment to the tidyverse!</strong></p>
<p>Website: <a href="https://stemangiola.github.io/tidySingleCellExperiment/articles/introduction.html" class="external-link">tidySingleCellExperiment</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidySummarizedExperiment/" class="external-link">tidySummarizedExperiment</a> for tidy manipulation of SummarizedExperiment objects)</li>
<li>
<a href="https://stemangiola.github.io/tidyseurat/" class="external-link">tidyseurat</a> for tidy manipulation of Seurat objects</li>
<li>
<a href="https://stemangiola.github.io/tidybulk/" class="external-link">tidybulk</a> for tidy bulk RNA-seq data analysis</li>
<li>
<a href="https://github.com/stemangiola/tidygate" class="external-link">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/" class="external-link">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>tidySingleCellExperiment provides a bridge between Bioconductor single-cell packages [@amezquita2019orchestrating] and the tidyverse [@wickham2019welcome]. It enables viewing the Bioconductor <em>SingleCellExperiment</em> object as a tidyverse tibble, and provides SingleCellExperiment-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions. This allows users to get the best of both Bioconductor and tidyverse worlds.</p>
<div class="section level2">
<h2 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h2>
<table class="table">
<colgroup>
<col width="33%">
<col width="66%">
</colgroup>
<thead><tr class="header">
<th>SingleCellExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySingleCellExperiment</code> is a SingleCellExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<colgroup>
<col width="27%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> tibble functions (e.g. <code>select</code>)</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> tibble functions (e.g. <code>pivot_longer</code>)</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> (<code>ggplot</code>)</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> (<code>plot_ly</code>)</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="22%">
<col width="77%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="even">
<td><code>join_features</code></td>
<td>Add feature-wise information, returns a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>aggregate_cells</code></td>
<td>Aggregate cell gene-transcription abundance as pseudobulk tissue</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidySingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<p>Load libraries used in this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bioconductor single-cell packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellSignalR</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Tidyverse-compatible packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org" class="external-link">tidyHeatmap</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Both</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="data-representation-of-tidysinglecellexperiment">Data representation of <code>tidySingleCellExperiment</code>
<a class="anchor" aria-label="anchor" href="#data-representation-of-tidysinglecellexperiment"></a>
</h1>
<p>This is a <em>SingleCellExperiment</em> object but it is evaluated as a tibble. So it is compatible both with SingleCellExperiment and tidyverse.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbmc_small</span>, package<span class="op">=</span><span class="st">"tidySingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">## [90m# A SingleCellExperiment-tibble abstraction: 80 × 17[39m</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="do">## [90m# [90mFeatures=230 | Cells=80 | Assays=counts, logcounts[0m[39m</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="do">##    .cell orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;fct&gt;[39m[23m         [3m[90m&lt;chr&gt;[39m[23m </span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="do">## [90m 1[39m ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="do">## [90m 2[39m CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="do">## [90m 3[39m GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="do">## [90m 4[39m TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="do">## [90m 5[39m AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="do">## [90m 6[39m TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="do">## [90m 7[39m TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="do">## [90m 8[39m GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="do">## [90m 9[39m GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="do">## [90m10[39m AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="do">## [90m# ℹ 70 more rows[39m</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="do">## [90m# ℹ 10 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;,[39m</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="do">## [90m#   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,[39m</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="do">## [90m#   tSNE_2 &lt;dbl&gt;[39m</span></span></code></pre></div>
<p><strong>But it is a SingleCellExperiment object after all</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="do">##         ATGCCAGAACGACT CATGGCCTGTGCAT GAACCTGATGAACC TGACTGGATTCTCA</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="do">## MS4A1                .              .              .              .</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="do">## CD79B                1              .              .              .</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="do">## CD79A                .              .              .              .</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="do">## HLA-DRA              .              1              .              .</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="do">## TCL1A                .              .              .              .</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="do">##         AGTCAGACTGCACA</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="do">## MS4A1                .</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="do">## CD79B                .</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="do">## CD79A                .</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="do">## HLA-DRA              1</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="do">## TCL1A                .</span></span></code></pre></div>
<p>The <code>SingleCellExperiment</code> object’s tibble visualisation can be turned off, or back on at any time.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Turn off the tibble visualisation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmc_small</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="do">## class: SingleCellExperiment </span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="do">## dim: 230 80 </span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="do">## metadata(0):</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="do">## assays(2): counts logcounts</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="do">## rownames(230): MS4A1 CD79B ... SPON2 S100B</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="do">## rowData names(5): vst.mean vst.variance vst.variance.expected</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="do">##   vst.variance.standardized vst.variable</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="do">## colnames(80): ATGCCAGAACGACT CATGGCCTGTGCAT ... GGAACACTTCAGAC</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="do">##   CTTGATTGATCTTC</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="do">## colData names(9): orig.ident nCount_RNA ... file ident</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Turn on the tibble visualisation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_SingleCellExperiment_show"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="annotation-polishing">Annotation polishing<a class="anchor" aria-label="anchor" href="#annotation-polishing"></a>
</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span><span class="op">$</span><span class="va">file</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## [1] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="do">## [2] "../data/sample1/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="do">## [3] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="do">## [4] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="do">## [5] "../data/sample2/outs/filtered_feature_bc_matrix/"</span></span></code></pre></div>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create sample column</span></span>
<span><span class="va">pbmc_small_polished</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/extract.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, remove<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reorder to have sample column up front</span></span>
<span><span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="do">## [90m# A SingleCellExperiment-tibble abstraction: 80 × 18[39m</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="do">## [90m# [90mFeatures=230 | Cells=80 | Assays=counts, logcounts[0m[39m</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="do">##    .cell sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m [3m[90m&lt;chr&gt;[39m[23m  [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;fct&gt;[39m[23m        </span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="do">## [90m 1[39m ATGC… sampl… SeuratPro…         70           47 0               A            </span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="do">## [90m 2[39m CATG… sampl… SeuratPro…         85           52 0               A            </span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="do">## [90m 3[39m GAAC… sampl… SeuratPro…         87           50 1               B            </span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="do">## [90m 4[39m TGAC… sampl… SeuratPro…        127           56 0               A            </span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="do">## [90m 5[39m AGTC… sampl… SeuratPro…        173           53 0               A            </span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="do">## [90m 6[39m TCTG… sampl… SeuratPro…         70           48 0               A            </span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="do">## [90m 7[39m TGGT… sampl… SeuratPro…         64           36 0               A            </span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="do">## [90m 8[39m GCAG… sampl… SeuratPro…         72           45 0               A            </span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="do">## [90m 9[39m GATA… sampl… SeuratPro…         52           36 0               A            </span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="do">## [90m10[39m AATG… sampl… SeuratPro…        100           41 0               A            </span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="do">## [90m# ℹ 70 more rows[39m</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="do">## [90m# ℹ 11 more variables: groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,[39m</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="do">## [90m#   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,[39m</span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="do">## [90m#   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;[39m</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="preliminary-plots">Preliminary plots<a class="anchor" aria-label="anchor" href="#preliminary-plots"></a>
</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use colourblind-friendly colours</span></span>
<span><span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html" class="external-link">dittoColors</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set theme</span></span>
<span><span class="va">custom_theme</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">friendly_cols</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">friendly_cols</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>                panel.border<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                axis.line<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                panel.grid.major<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>,</span>
<span>                panel.grid.minor<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>,</span>
<span>                text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,</span>
<span>                legend.position<span class="op">=</span><span class="st">"bottom"</span>,</span>
<span>                aspect.ratio<span class="op">=</span><span class="fl">1</span>,</span>
<span>                strip.background<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                axis.title.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                axis.title.y<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">## Warning: [1m[22mThe `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="do">## [36mℹ[39m Please use the `linewidth` argument instead.</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="do">## [90mThis warning is displayed once every 8 hours.[39m</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="do">## [90mCall `lifecycle::last_lifecycle_warnings()` to see where this warning was[39m</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="do">## [90mgenerated.[39m</span></span></code></pre></div>
<p>We can treat <code>pbmc_small_polished</code> as a tibble for plotting.</p>
<p>Here we plot number of features per cell.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">nFeature_RNA</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="do">## [1m[22m`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/plot1-1.png"><!-- --></p>
<p>Here we plot total features per cell.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">nCount_RNA</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/plot2-1.png"><!-- --></p>
<p>Here we plot abundance of two features for each group.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">.abundance_counts</span> <span class="op">+</span> <span class="fl">1</span>, fill<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">nCount_RNA</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, width<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: join_features produces duplicate cell names to accomadate the long data format. For this reason, a data frame is returned for independent data analysis. Assay feature abundance is appended as .abundance_counts and .abundance_logcounts.</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/unnamed-chunk-12-1.png"><!-- --></p>
</div>
<div class="section level1">
<h1 id="preprocess-the-dataset">Preprocess the dataset<a class="anchor" aria-label="anchor" href="#preprocess-the-dataset"></a>
</h1>
<p>We can also treat <code>pbmc_small_polished</code> as a <em>SingleCellExperiment</em> object and proceed with data processing with Bioconductor packages, such as <em>scran</em> [@lun2016pooling] and <em>scater</em> [@mccarthy2017scater].</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify variable genes with scran</span></span>
<span><span class="va">variable_genes</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform PCA with scater</span></span>
<span><span class="va">pbmc_small_pca</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_polished</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span>subset_row<span class="op">=</span><span class="va">variable_genes</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="do">## Warning in check_numbers(k = k, nu = nu, nv = nv, limit = min(dim(x)) - : more</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="do">## singular values/vectors requested than available</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a><span class="do">## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="do">## TRUE, : You're computing too large a percentage of total singular values, use a</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a><span class="do">## standard svd instead.</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_pca</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="do">## [90m# A SingleCellExperiment-tibble abstraction: 80 × 18[39m</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="do">## [90m# [90mFeatures=230 | Cells=80 | Assays=counts, logcounts[0m[39m</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="do">##    .cell orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;fct&gt;[39m[23m         [3m[90m&lt;chr&gt;[39m[23m </span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a><span class="do">## [90m 1[39m ATGC… SeuratPro…         70           47 0               A             g2    </span></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a><span class="do">## [90m 2[39m CATG… SeuratPro…         85           52 0               A             g1    </span></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a><span class="do">## [90m 3[39m GAAC… SeuratPro…         87           50 1               B             g2    </span></span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a><span class="do">## [90m 4[39m TGAC… SeuratPro…        127           56 0               A             g2    </span></span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a><span class="do">## [90m 5[39m AGTC… SeuratPro…        173           53 0               A             g2    </span></span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a><span class="do">## [90m 6[39m TCTG… SeuratPro…         70           48 0               A             g1    </span></span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a><span class="do">## [90m 7[39m TGGT… SeuratPro…         64           36 0               A             g1    </span></span>
<span id="cb25-12"><a href="#cb25-12" tabindex="-1"></a><span class="do">## [90m 8[39m GCAG… SeuratPro…         72           45 0               A             g1    </span></span>
<span id="cb25-13"><a href="#cb25-13" tabindex="-1"></a><span class="do">## [90m 9[39m GATA… SeuratPro…         52           36 0               A             g1    </span></span>
<span id="cb25-14"><a href="#cb25-14" tabindex="-1"></a><span class="do">## [90m10[39m AATG… SeuratPro…        100           41 0               A             g1    </span></span>
<span id="cb25-15"><a href="#cb25-15" tabindex="-1"></a><span class="do">## [90m# ℹ 70 more rows[39m</span></span>
<span id="cb25-16"><a href="#cb25-16" tabindex="-1"></a><span class="do">## [90m# ℹ 11 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;,[39m</span></span>
<span id="cb25-17"><a href="#cb25-17" tabindex="-1"></a><span class="do">## [90m#   ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;,[39m</span></span>
<span id="cb25-18"><a href="#cb25-18" tabindex="-1"></a><span class="do">## [90m#   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;[39m</span></span></code></pre></div>
<p>If a tidyverse-compatible package is not included in the tidySingleCellExperiment collection, we can use <code>as_tibble</code> to permanently convert <code>tidySingleCellExperiment</code> into a tibble.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create pairs plot with GGally</span></span>
<span><span class="va">pbmc_small_pca</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"PC"</span><span class="op">)</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="do">## Registered S3 method overwritten by 'GGally':</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="do">##   method from   </span></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a><span class="do">##   +.gg   ggplot2</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/pc_plot-1.png"><!-- --></p>
</div>
<div class="section level1">
<h1 id="identify-clusters">Identify clusters<a class="anchor" aria-label="anchor" href="#identify-clusters"></a>
</h1>
<p>We can proceed with cluster identification with <em>scran</em>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cluster</span> <span class="op">&lt;-</span> <span class="va">pbmc_small_pca</span></span>
<span></span>
<span><span class="co"># Assign clusters to the 'colLabels' of the SingleCellExperiment object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">pbmc_small_cluster</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_pca</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/exposition.html" class="external-link">%$%</a></span></span>
<span>    <span class="va">membership</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="do">## Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="do">## detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reorder columns</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="do">## [90m# A SingleCellExperiment-tibble abstraction: 80 × 19[39m</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="do">## [90m# [90mFeatures=230 | Cells=80 | Assays=counts, logcounts[0m[39m</span></span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="do">##    .cell  label orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents</span></span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m  [3m[90m&lt;fct&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;fct&gt;[39m[23m        </span></span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="do">## [90m 1[39m ATGCC… 2     SeuratPro…         70           47 0               A            </span></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a><span class="do">## [90m 2[39m CATGG… 2     SeuratPro…         85           52 0               A            </span></span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a><span class="do">## [90m 3[39m GAACC… 2     SeuratPro…         87           50 1               B            </span></span>
<span id="cb31-8"><a href="#cb31-8" tabindex="-1"></a><span class="do">## [90m 4[39m TGACT… 1     SeuratPro…        127           56 0               A            </span></span>
<span id="cb31-9"><a href="#cb31-9" tabindex="-1"></a><span class="do">## [90m 5[39m AGTCA… 2     SeuratPro…        173           53 0               A            </span></span>
<span id="cb31-10"><a href="#cb31-10" tabindex="-1"></a><span class="do">## [90m 6[39m TCTGA… 2     SeuratPro…         70           48 0               A            </span></span>
<span id="cb31-11"><a href="#cb31-11" tabindex="-1"></a><span class="do">## [90m 7[39m TGGTA… 1     SeuratPro…         64           36 0               A            </span></span>
<span id="cb31-12"><a href="#cb31-12" tabindex="-1"></a><span class="do">## [90m 8[39m GCAGC… 2     SeuratPro…         72           45 0               A            </span></span>
<span id="cb31-13"><a href="#cb31-13" tabindex="-1"></a><span class="do">## [90m 9[39m GATAT… 2     SeuratPro…         52           36 0               A            </span></span>
<span id="cb31-14"><a href="#cb31-14" tabindex="-1"></a><span class="do">## [90m10[39m AATGT… 2     SeuratPro…        100           41 0               A            </span></span>
<span id="cb31-15"><a href="#cb31-15" tabindex="-1"></a><span class="do">## [90m# ℹ 70 more rows[39m</span></span>
<span id="cb31-16"><a href="#cb31-16" tabindex="-1"></a><span class="do">## [90m# ℹ 12 more variables: groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,[39m</span></span>
<span id="cb31-17"><a href="#cb31-17" tabindex="-1"></a><span class="do">## [90m#   sample &lt;chr&gt;, ident &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;,[39m</span></span>
<span id="cb31-18"><a href="#cb31-18" tabindex="-1"></a><span class="do">## [90m#   PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;[39m</span></span></code></pre></div>
<p>And interrogate the output as if it was a regular tibble.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count number of cells for each cluster per group</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/count.html">count</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a><span class="do">## [90m# A tibble: 8 × 3[39m</span></span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a><span class="do">##   groups label     n</span></span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a><span class="do">##   [3m[90m&lt;chr&gt;[39m[23m  [3m[90m&lt;fct&gt;[39m[23m [3m[90m&lt;int&gt;[39m[23m</span></span>
<span id="cb33-6"><a href="#cb33-6" tabindex="-1"></a><span class="do">## [90m1[39m g1     1        12</span></span>
<span id="cb33-7"><a href="#cb33-7" tabindex="-1"></a><span class="do">## [90m2[39m g1     2        14</span></span>
<span id="cb33-8"><a href="#cb33-8" tabindex="-1"></a><span class="do">## [90m3[39m g1     3        14</span></span>
<span id="cb33-9"><a href="#cb33-9" tabindex="-1"></a><span class="do">## [90m4[39m g1     4         4</span></span>
<span id="cb33-10"><a href="#cb33-10" tabindex="-1"></a><span class="do">## [90m5[39m g2     1        10</span></span>
<span id="cb33-11"><a href="#cb33-11" tabindex="-1"></a><span class="do">## [90m6[39m g2     2        11</span></span>
<span id="cb33-12"><a href="#cb33-12" tabindex="-1"></a><span class="do">## [90m7[39m g2     3        10</span></span>
<span id="cb33-13"><a href="#cb33-13" tabindex="-1"></a><span class="do">## [90m8[39m g2     4         5</span></span></code></pre></div>
<p>We can identify and visualise cluster markers combining SingleCellExperiment, tidyverse functions and tidyHeatmap [@mangiola2020tidyheatmap]</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify top 10 markers per cluster</span></span>
<span><span class="va">marker_genes</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cluster</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span>groups<span class="op">=</span><span class="va">pbmc_small_cluster</span><span class="op">$</span><span class="va">label</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot heatmap</span></span>
<span><span class="va">pbmc_small_cluster</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="va">marker_genes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-method.html" class="external-link">heatmap</a></span><span class="op">(</span><span class="va">.feature</span>, <span class="va">.cell</span>, <span class="va">.abundance_counts</span>, .scale<span class="op">=</span><span class="st">"column"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: join_features produces duplicate cell names to accomadate the long data format. For this reason, a data frame is returned for independent data analysis. Assay feature abundance is appended as .abundance_counts and .abundance_logcounts.</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a><span class="do">## tidyHeatmap says: (once per session) from release 1.7.0 the scaling is set to "none" by default. Please use scale = "row", "column" or "both" to apply scaling</span></span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a></span>
<span id="cb35-5"><a href="#cb35-5" tabindex="-1"></a><span class="do">## Warning: [1m[22mThe `.scale` argument of `heatmap()` is deprecated as of tidyHeatmap 1.7.0.</span></span>
<span id="cb35-6"><a href="#cb35-6" tabindex="-1"></a><span class="do">## [36mℹ[39m Please use scale (without dot prefix) instead: heatmap(scale = ...)</span></span>
<span id="cb35-7"><a href="#cb35-7" tabindex="-1"></a><span class="do">## [90mThis warning is displayed once every 8 hours.[39m</span></span>
<span id="cb35-8"><a href="#cb35-8" tabindex="-1"></a><span class="do">## [90mCall `lifecycle::last_lifecycle_warnings()` to see where this warning was[39m</span></span>
<span id="cb35-9"><a href="#cb35-9" tabindex="-1"></a><span class="do">## [90mgenerated.[39m</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/unnamed-chunk-13-1.png"><!-- --></p>
</div>
<div class="section level1">
<h1 id="reduce-dimensions">Reduce dimensions<a class="anchor" aria-label="anchor" href="#reduce-dimensions"></a>
</h1>
<p>We can calculate the first 3 UMAP dimensions using the SingleCellExperiment framework and <em>scater</em>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_UMAP</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cluster</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>ncomponents<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>And we can plot the result in 3D using plotly.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_UMAP</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span><span class="op">(</span></span>
<span>        x<span class="op">=</span><span class="op">~</span><span class="va">`UMAP1`</span>,</span>
<span>        y<span class="op">=</span><span class="op">~</span><span class="va">`UMAP2`</span>,</span>
<span>        z<span class="op">=</span><span class="op">~</span><span class="va">`UMAP3`</span>,</span>
<span>        color<span class="op">=</span><span class="op">~</span><span class="va">label</span>,</span>
<span>        colors<span class="op">=</span><span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<figure><img src="inst/extdata/plotly.png" alt="plotly screenshot"><figcaption aria-hidden="true">
plotly screenshot
</figcaption></figure>
</div>
<div class="section level1">
<h1 id="cell-type-prediction">Cell type prediction<a class="anchor" aria-label="anchor" href="#cell-type-prediction"></a>
</h1>
<p>We can infer cell type identities using <em>SingleR</em> [@aran2019reference] and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get cell type reference data</span></span>
<span><span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/BlueprintEncodeData.html" class="external-link">BlueprintEncodeData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Infer cell identities</span></span>
<span><span class="va">cell_type_df</span> <span class="op">&lt;-</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pbmc_small_UMAP</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span>sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">SingleR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span></span>
<span>        ref <span class="op">=</span> <span class="va">blueprint</span>,</span>
<span>        labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"single"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"cell"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Join UMAP and cell type info</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cell_type_df</span><span class="op">)</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_UMAP</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/left_join.html">left_join</a></span><span class="op">(</span><span class="va">cell_type_df</span>, by<span class="op">=</span><span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="do">## Warning in is_sample_feature_deprecated_used(x, .cols):</span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: from version 1.3.1, the special columns</span></span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a><span class="do">## including cell id (colnames(se)) has changed to ".cell". This dataset is</span></span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a><span class="do">## returned with the old-style vocabulary (cell), however, we suggest to update</span></span>
<span id="cb40-5"><a href="#cb40-5" tabindex="-1"></a><span class="do">## your workflow to reflect the new vocabulary (.cell).</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reorder columns</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="do">## Warning in is_sample_feature_deprecated_used(.data, .cols):</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: from version 1.3.1, the special columns</span></span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a><span class="do">## including cell id (colnames(se)) has changed to ".cell". This dataset is</span></span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="do">## returned with the old-style vocabulary (cell), however, we suggest to update</span></span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a><span class="do">## your workflow to reflect the new vocabulary (.cell).</span></span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" tabindex="-1"></a><span class="do">## [90m# A SingleCellExperiment-tibble abstraction: 80 × 23[39m</span></span>
<span id="cb42-8"><a href="#cb42-8" tabindex="-1"></a><span class="do">## [90m# [90mFeatures=230 | Cells=80 | Assays=counts, logcounts[0m[39m</span></span>
<span id="cb42-9"><a href="#cb42-9" tabindex="-1"></a><span class="do">##    cell          first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8</span></span>
<span id="cb42-10"><a href="#cb42-10" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m         [3m[90m&lt;chr&gt;[39m[23m        [3m[90m&lt;fct&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;fct&gt;[39m[23m          </span></span>
<span id="cb42-11"><a href="#cb42-11" tabindex="-1"></a><span class="do">## [90m 1[39m ATGCCAGAACGA… CD4+ T-cells SeuratPro…         70           47 0              </span></span>
<span id="cb42-12"><a href="#cb42-12" tabindex="-1"></a><span class="do">## [90m 2[39m CATGGCCTGTGC… CD8+ T-cells SeuratPro…         85           52 0              </span></span>
<span id="cb42-13"><a href="#cb42-13" tabindex="-1"></a><span class="do">## [90m 3[39m GAACCTGATGAA… CD8+ T-cells SeuratPro…         87           50 1              </span></span>
<span id="cb42-14"><a href="#cb42-14" tabindex="-1"></a><span class="do">## [90m 4[39m TGACTGGATTCT… CD4+ T-cells SeuratPro…        127           56 0              </span></span>
<span id="cb42-15"><a href="#cb42-15" tabindex="-1"></a><span class="do">## [90m 5[39m AGTCAGACTGCA… CD4+ T-cells SeuratPro…        173           53 0              </span></span>
<span id="cb42-16"><a href="#cb42-16" tabindex="-1"></a><span class="do">## [90m 6[39m TCTGATACACGT… CD4+ T-cells SeuratPro…         70           48 0              </span></span>
<span id="cb42-17"><a href="#cb42-17" tabindex="-1"></a><span class="do">## [90m 7[39m TGGTATCTAAAC… CD4+ T-cells SeuratPro…         64           36 0              </span></span>
<span id="cb42-18"><a href="#cb42-18" tabindex="-1"></a><span class="do">## [90m 8[39m GCAGCTCTGTTT… CD4+ T-cells SeuratPro…         72           45 0              </span></span>
<span id="cb42-19"><a href="#cb42-19" tabindex="-1"></a><span class="do">## [90m 9[39m GATATAACACGC… CD4+ T-cells SeuratPro…         52           36 0              </span></span>
<span id="cb42-20"><a href="#cb42-20" tabindex="-1"></a><span class="do">## [90m10[39m AATGTTGACAGT… CD4+ T-cells SeuratPro…        100           41 0              </span></span>
<span id="cb42-21"><a href="#cb42-21" tabindex="-1"></a><span class="do">## [90m# ℹ 70 more rows[39m</span></span>
<span id="cb42-22"><a href="#cb42-22" tabindex="-1"></a><span class="do">## [90m# ℹ 17 more variables: letter.idents &lt;fct&gt;, groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;,[39m</span></span>
<span id="cb42-23"><a href="#cb42-23" tabindex="-1"></a><span class="do">## [90m#   file &lt;chr&gt;, sample &lt;chr&gt;, ident &lt;fct&gt;, label &lt;fct&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,[39m</span></span>
<span id="cb42-24"><a href="#cb42-24" tabindex="-1"></a><span class="do">## [90m#   PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;,[39m</span></span>
<span id="cb42-25"><a href="#cb42-25" tabindex="-1"></a><span class="do">## [90m#   UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;[39m</span></span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count number of cells for each cell type per cluster</span></span>
<span><span class="va">pbmc_small_cell_type</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/count.html">count</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a><span class="do">## [90m# A tibble: 11 × 3[39m</span></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="do">##    label first.labels     n</span></span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a><span class="do">##    [3m[90m&lt;fct&gt;[39m[23m [3m[90m&lt;chr&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m</span></span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a><span class="do">## [90m 1[39m 1     CD4+ T-cells     2</span></span>
<span id="cb44-7"><a href="#cb44-7" tabindex="-1"></a><span class="do">## [90m 2[39m 1     CD8+ T-cells     8</span></span>
<span id="cb44-8"><a href="#cb44-8" tabindex="-1"></a><span class="do">## [90m 3[39m 1     NK cells        12</span></span>
<span id="cb44-9"><a href="#cb44-9" tabindex="-1"></a><span class="do">## [90m 4[39m 2     B-cells         10</span></span>
<span id="cb44-10"><a href="#cb44-10" tabindex="-1"></a><span class="do">## [90m 5[39m 2     CD4+ T-cells     6</span></span>
<span id="cb44-11"><a href="#cb44-11" tabindex="-1"></a><span class="do">## [90m 6[39m 2     CD8+ T-cells     2</span></span>
<span id="cb44-12"><a href="#cb44-12" tabindex="-1"></a><span class="do">## [90m 7[39m 2     Macrophages      1</span></span>
<span id="cb44-13"><a href="#cb44-13" tabindex="-1"></a><span class="do">## [90m 8[39m 2     Monocytes        6</span></span>
<span id="cb44-14"><a href="#cb44-14" tabindex="-1"></a><span class="do">## [90m 9[39m 3     Macrophages      1</span></span>
<span id="cb44-15"><a href="#cb44-15" tabindex="-1"></a><span class="do">## [90m10[39m 3     Monocytes       23</span></span>
<span id="cb44-16"><a href="#cb44-16" tabindex="-1"></a><span class="do">## [90m11[39m 4     Erythrocytes     9</span></span></code></pre></div>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cell_type</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Reshape and add classifier column</span></span>
<span>    <span class="fu"><a href="reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span></span>
<span>        cols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span>,</span>
<span>        names_to<span class="op">=</span><span class="st">"classifier"</span>, values_to<span class="op">=</span><span class="st">"label"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># UMAP plots for cell type and cluster</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>, <span class="va">UMAP2</span>, color<span class="op">=</span><span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">classifier</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/unnamed-chunk-17-1.png"><!-- --></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_cell_type</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Add some mitochondrial abundance values</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>mitochondrial<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Plot correlation</span></span>
<span>    <span class="fu"><a href="reference/join_features.html">join_features</a></span><span class="op">(</span>features<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CST3"</span>, <span class="st">"LYZ"</span><span class="op">)</span>, shape<span class="op">=</span><span class="st">"wide"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">CST3</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">LYZ</span> <span class="op">+</span> <span class="fl">1</span>, color<span class="op">=</span><span class="va">groups</span>, size<span class="op">=</span><span class="va">mitochondrial</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">first.labels</span>, scales<span class="op">=</span><span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="do">## Warning in is_sample_feature_deprecated_used(x, .cols):</span></span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a><span class="do">## tidySingleCellExperiment says: from version 1.3.1, the special columns</span></span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a><span class="do">## including cell id (colnames(se)) has changed to ".cell". This dataset is</span></span>
<span id="cb48-4"><a href="#cb48-4" tabindex="-1"></a><span class="do">## returned with the old-style vocabulary (cell), however, we suggest to update</span></span>
<span id="cb48-5"><a href="#cb48-5" tabindex="-1"></a><span class="do">## your workflow to reflect the new vocabulary (.cell).</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/unnamed-chunk-18-1.png"><!-- --></p>
</div>
<div class="section level1">
<h1 id="nested-analyses">Nested analyses<a class="anchor" aria-label="anchor" href="#nested-analyses"></a>
</h1>
<p>A powerful tool we can use with tidySingleCellExperiment is tidyverse <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types into lymphoid and myeloid, and then nest based on the new classification.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_cell_type</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">first.labels</span> <span class="op">!=</span> <span class="st">"Erythrocytes"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>cell_class<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">`first.labels`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span><span class="op">)</span>, <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/nest.html">nest</a></span><span class="op">(</span>data<span class="op">=</span><span class="op">-</span><span class="va">cell_class</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="do">## Warning: [1m[22mThere were 2 warnings in `mutate()`.</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a><span class="do">## The first warning was:</span></span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a><span class="do">## [1m[22m[36mℹ[39m In argument: `data = map(...)`.</span></span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a><span class="do">## Caused by warning in `is_sample_feature_deprecated_used()`:</span></span>
<span id="cb50-5"><a href="#cb50-5" tabindex="-1"></a><span class="do">## [33m![39m tidySingleCellExperiment says: from version 1.3.1, the special columns including cell id (colnames(se)) has changed to ".cell". This dataset is returned with the old-style vocabulary (cell), however, we suggest to update your workflow to reflect the new vocabulary (.cell).</span></span>
<span id="cb50-6"><a href="#cb50-6" tabindex="-1"></a><span class="do">## [1m[22m[36mℹ[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 2 × 2[39m</span></span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a><span class="do">##   cell_class data           </span></span>
<span id="cb52-3"><a href="#cb52-3" tabindex="-1"></a><span class="do">##   [3m[90m&lt;chr&gt;[39m[23m      [3m[90m&lt;list&gt;[39m[23m         </span></span>
<span id="cb52-4"><a href="#cb52-4" tabindex="-1"></a><span class="do">## [90m1[39m lymphoid   [90m&lt;SnglCllE[,40]&gt;[39m</span></span>
<span id="cb52-5"><a href="#cb52-5" tabindex="-1"></a><span class="do">## [90m2[39m myeloid    [90m&lt;SnglCllE[,31]&gt;[39m</span></span></code></pre></div>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_reanalysed</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_nested</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>        <span class="va">data</span>, <span class="op">~</span> <span class="op">{</span></span>
<span>            <span class="va">.x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">.x</span>, subset_row<span class="op">=</span><span class="va">variable_genes</span><span class="op">)</span></span>
<span></span>
<span>            <span class="va">variable_genes</span> <span class="op">&lt;-</span></span>
<span>                <span class="va">.x</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>prop<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>                <span class="va">.x</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/exposition.html" class="external-link">%$%</a></span></span>
<span>                <span class="va">membership</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>            <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>ncomponents<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_nested_reanalysed</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 2 × 2[39m</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="do">##   cell_class data           </span></span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="do">##   [3m[90m&lt;chr&gt;[39m[23m      [3m[90m&lt;list&gt;[39m[23m         </span></span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a><span class="do">## [90m1[39m lymphoid   [90m&lt;SnglCllE[,40]&gt;[39m</span></span>
<span id="cb54-5"><a href="#cb54-5" tabindex="-1"></a><span class="do">## [90m2[39m myeloid    [90m&lt;SnglCllE[,31]&gt;[39m</span></span></code></pre></div>
<p>We can then unnest and plot the new classification.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_reanalysed</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Define unique clusters</span></span>
<span>    <span class="fu"><a href="reference/unite.html">unite</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cell_class</span>, <span class="va">label</span><span class="op">)</span>, remove<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Plotting</span></span>
<span>    <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP1</span>, <span class="va">UMAP2</span>, color<span class="op">=</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">cell_class</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">custom_theme</span></span></code></pre></div>
<p><img src="inst/extdata/readme_figures/unnamed-chunk-21-1.png"><!-- --></p>
<p>We can perform a large number of functional analyses on data subsets. For example, we can identify intra-sample cell-cell interactions using <em>SingleCellSignalR</em> [@cabello2020singlecellsignalr], and then compare whether interactions are stronger or weaker across conditions. The code below demonstrates how this analysis could be performed. It won’t work with this small example dataset as we have just two samples (one for each condition). But some example output is shown below and you can imagine how you can use tidyverse on the output to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small_nested_interactions</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">pbmc_small_nested_reanalysed</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Unnest based on cell category</span></span>
<span>    <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Create unambiguous clusters</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>integrated_clusters<span class="op">=</span><span class="va">first.labels</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Nest based on sample</span></span>
<span>    <span class="fu"><a href="reference/nest.html">nest</a></span><span class="op">(</span>data<span class="op">=</span><span class="op">-</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>interactions<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="op">{</span></span>
<span></span>
<span>        <span class="co"># Produce variables. Yuck!</span></span>
<span>        <span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">$</span><span class="va">integrated_clusters</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract2</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span></span>
<span>        <span class="va">data</span> <span class="op">|&gt;</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/cell_signaling.html" class="external-link">cell_signaling</a></span><span class="op">(</span>genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, cluster<span class="op">=</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/inter_network.html" class="external-link">inter_network</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span>, signal<span class="op">=</span><span class="va">_</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, cluster<span class="op">=</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/exposition.html" class="external-link">%$%</a></span></span>
<span>            <span class="va">`individual-networks`</span> <span class="op">|&gt;</span></span>
<span>            <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_small_nested_interactions</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">interactions</span><span class="op">)</span></span></code></pre></div>
<p>If the dataset was not so small, and interactions could be identified, you would see something like below.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbmc_small_nested_interactions</span><span class="op">)</span></span>
<span><span class="va">pbmc_small_nested_interactions</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 100 × 9[39m</span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a><span class="do">##    sample  ligand          receptor ligand.name receptor.name origin destination</span></span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m   [3m[90m&lt;chr&gt;[39m[23m           [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;chr&gt;[39m[23m       [3m[90m&lt;chr&gt;[39m[23m         [3m[90m&lt;chr&gt;[39m[23m  [3m[90m&lt;chr&gt;[39m[23m      </span></span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a><span class="do">## [90m 1[39m sample1 cluster 1.PTMA  cluster… PTMA        VIPR1         clust… cluster 2  </span></span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a><span class="do">## [90m 2[39m sample1 cluster 1.B2M   cluster… B2M         KLRD1         clust… cluster 2  </span></span>
<span id="cb58-6"><a href="#cb58-6" tabindex="-1"></a><span class="do">## [90m 3[39m sample1 cluster 1.IL16  cluster… IL16        CD4           clust… cluster 2  </span></span>
<span id="cb58-7"><a href="#cb58-7" tabindex="-1"></a><span class="do">## [90m 4[39m sample1 cluster 1.HLA-B cluster… HLA-B       KLRD1         clust… cluster 2  </span></span>
<span id="cb58-8"><a href="#cb58-8" tabindex="-1"></a><span class="do">## [90m 5[39m sample1 cluster 1.CALM1 cluster… CALM1       VIPR1         clust… cluster 2  </span></span>
<span id="cb58-9"><a href="#cb58-9" tabindex="-1"></a><span class="do">## [90m 6[39m sample1 cluster 1.HLA-E cluster… HLA-E       KLRD1         clust… cluster 2  </span></span>
<span id="cb58-10"><a href="#cb58-10" tabindex="-1"></a><span class="do">## [90m 7[39m sample1 cluster 1.GNAS  cluster… GNAS        VIPR1         clust… cluster 2  </span></span>
<span id="cb58-11"><a href="#cb58-11" tabindex="-1"></a><span class="do">## [90m 8[39m sample1 cluster 1.B2M   cluster… B2M         HFE           clust… cluster 2  </span></span>
<span id="cb58-12"><a href="#cb58-12" tabindex="-1"></a><span class="do">## [90m 9[39m sample1 cluster 1.PTMA  cluster… PTMA        VIPR1         clust… cluster 3  </span></span>
<span id="cb58-13"><a href="#cb58-13" tabindex="-1"></a><span class="do">## [90m10[39m sample1 cluster 1.CALM1 cluster… CALM1       VIPR1         clust… cluster 3  </span></span>
<span id="cb58-14"><a href="#cb58-14" tabindex="-1"></a><span class="do">## [90m# ℹ 90 more rows[39m</span></span>
<span id="cb58-15"><a href="#cb58-15" tabindex="-1"></a><span class="do">## [90m# ℹ 2 more variables: interaction.type &lt;chr&gt;, LRscore &lt;dbl&gt;[39m</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="aggregating-cells">Aggregating cells<a class="anchor" aria-label="anchor" href="#aggregating-cells"></a>
</h1>
<p>Sometimes, it is necessary to aggregate the gene-transcript abundance from a group of cells into a single value. For example, when comparing groups of cells across different samples with fixed-effect models.</p>
<p>In tidySingleCellExperiment, cell aggregation can be achieved using the <code>aggregate_cells</code> function.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/aggregate_cells.html">aggregate_cells</a></span><span class="op">(</span><span class="va">groups</span>, assays <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="do">## class: SummarizedExperiment </span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a><span class="do">## dim: 230 2 </span></span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a><span class="do">## metadata(0):</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a><span class="do">## assays(1): counts</span></span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a><span class="do">## rownames(230): ACAP1 ACRBP ... ZNF330 ZNF76</span></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a><span class="do">## rowData names(0):</span></span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a><span class="do">## colnames(2): g1 g2</span></span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a><span class="do">## colData names(4): .aggregated_cells groups orig.ident file</span></span></code></pre></div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/tidySingleCellExperiment" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/stemangiola/tidySingleCellExperiment/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/stemangiola/tidySingleCellExperiment/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidySingleCellExperiment</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-7474-836X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidySingleCellExperiment/actions" class="external-link"><img src="https://github.com/stemangiola/tidySingleCellExperiment/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
